eo-phi-normalizer/app/Main.hs

```diff
@@ -1,111 +1,222 @@
+  deriving (Show)
+
+_FILE :: String
+_FILE = "FILE"
+
+_PROGRAM :: String
+_PROGRAM = "PROGRAM"
+
+fileMetavar :: Mod OptionFields a
+fileMetavar = metavar _FILE
+
+inputFileLongName :: String
+inputFileLongName = "input-file"
+
+inputFileOption :: Parser (Maybe FilePath)
+inputFileOption = optional $ strOption (long inputFileLongName <> short 'i' <> help [i|#{_FILE} to read input from. When #{_FILE} is -, read from stdin. You must specify either this option or #{_PROGRAM}.|] <> fileMetavar)
+
+outputFileOption :: Parser (Maybe String)
+outputFileOption = optional $ strOption (long "output-file" <> short 'o' <> help [i|Output to #{_FILE}. Output to stdout otherwise.|] <> fileMetavar)
+
+programArg :: Parser (Maybe String)
+programArg = optional $ strArgument (metavar _PROGRAM <> help "Program to work with.")
+
+jsonSwitch :: Parser Bool
+jsonSwitch = switch (long "json" <> short 'j' <> help "Output JSON.")
+
+cli'TransformPhi :: Parser CLI'TransformPhi
+cli'TransformPhi = do
+  rulesPath <- strOption (long "rules" <> short 'r' <> help [i|#{_FILE} with user-defined rules.|] <> fileMetavar)
+  inputFile <- inputFileOption
+  program <- programArg
+  chain <- switch (long "chain" <> short 'c' <> help "Output transformation steps.")
+  json <- jsonSwitch
+  outputFile <- outputFileOption
+  single <- switch (long "single" <> short 's' <> help "Output a single expression.")
+  pure CLI'TransformPhi{..}
 
-data CLIOptions = CLIOptions CLINamedParams (Maybe FilePath)
-  deriving (Generic, Show, ParseRecord)
+cli'MetricsPhi :: Parser CLI'MetricsPhi
+cli'MetricsPhi = do
+  inputFile <- inputFileOption
+  outputFile <- outputFileOption
+  program <- programArg
+  pure CLI'MetricsPhi{..}
+
+metricsParserInfo :: ParserInfo CLI
+metricsParserInfo = info (CLI'MetricsPhi' <$> cli'MetricsPhi) (progDesc "Collect metrics for a PHI program.")
+
+transformParserInfo :: ParserInfo CLI
+transformParserInfo = info (CLI'TransformPhi' <$> cli'TransformPhi) (progDesc "Transform a PHI program.")
+
+transformCommandName :: String
+transformCommandName = "transform"
+
+metricsCommandName :: String
+metricsCommandName = "metrics"
+
+cli :: Parser CLI
+cli =
+  hsubparser
+    ( command transformCommandName transformParserInfo
+        <> command metricsCommandName metricsParserInfo
+    )
+
+cliOpts :: ParserInfo CLI
+cliOpts =
+  info
+    (cli <**> helper)
+    (fullDesc <> progDesc "Work with PHI expressions.")
 
 data StructuredJSON = StructuredJSON
   { input :: String
-  , results :: [[String]]
+  , output :: [[String]]
   }
   deriving (Generic, ToJSON)
 
-encodeString :: (ToJSON a) => a -> String
-encodeString = (^. unpacked) . encodeToLazyText
+encodeToJSONString :: (ToJSON a) => a -> String
+encodeToJSONString = (^. unpacked) . toLazyText . encodePrettyToTextBuilder' defConfig{confIndent = Spaces 2}
+
+pprefs :: ParserPrefs
+pprefs = prefs (showHelpOnEmpty <> showHelpOnError)
+
+type Context = (?parserContext :: Optparse.Context)
```

Comment:
I'm not familiar with many extensions, what does this syntax mean?

Reply:
It's [ImplicitParams](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/implicit_parameters.html).

I use this extension instead of the [ReaderT pattern](https://www.fpcomplete.com/blog/readert-design-pattern/).